src/components/TreeView.js

```diff
@@ -4,20 +4,26 @@
 import React from 'react';
 import _ from 'lodash';
 import {EasySelect} from './EasySelect';
+import SIGNATURE from '../constants/signature';
 
 // @param {array} props.nodes - Array of TreeView compatible nodes
 export default class TreeView extends React.Component {
   render() {
-    let {nodes, className} = this.props;
+    let {nodes, className, fetchedSigners} = this.props;
     let rootClass = 'TreeView ' + (className) ? className : '';
 
+    // First check in case of transaction with no signatures, i.e. pre-auth
+    if(nodes[0] && nodes[0].type == "hint" && fetchedSigners) {
```

Comment:
I think this condition is a little complicated. `fetchedSigners` is present if the type is a `TransactionEnvelope` (checked in `state.type === "TransactionEnvelope"` condition). So we really want to check if we are in `TransactionEnvelope` view but we're doing it by checking the other variable set somewhere else.

Also I believe that in a very near future we'll want to fix https://github.com/stellar/laboratory/issues/336 and it will be much easier to do if we know the exact location we're currently at in a tree.

So I propose to construct a location string. We always have key name in `nodes[i].type`. So we can construct a location string like `TransactionEnvelope.signatures[i]` by appending the current key name in each child and trigger an action when we are when we want to be. For location with a variable part ( arrays like `TransactionEnvelope.signatures[0]`, `TransactionEnvelope.signatures[1]`...) we can use `RegExp`.

Reply:
Not quite following here -- can discuss tomorrow AM (UTC -7).

I see the key name, but where does the location string live? Is it attached to the node? I know that the TreeView has some recursion, so does this key persist through iterations?

Reply:
We need to build a location ourselves. `TreeView` renders itself recursively so if we send a current (then parent) name to the child we can construct the full location string. Check the quick patch below:

```diff
diff --git a/src/components/TreeView.js b/src/components/TreeView.js
index e7ed97b..8a0e4d4 100644
--- a/src/components/TreeView.js
+++ b/src/components/TreeView.js
@@ -9,7 +9,7 @@ import SIGNATURE from '../constants/signature';
 // @param {array} props.nodes - Array of TreeView compatible nodes
 export default class TreeView extends React.Component {
   render() {
-    let {nodes, className, fetchedSigners} = this.props;
+    let {nodes, className, fetchedSigners, parent} = this.props;
     let rootClass = 'TreeView ' + (className) ? className : '';
 
     if(nodes[1] && nodes[1].type == "signatures") {
@@ -20,15 +20,22 @@ export default class TreeView extends React.Component {
       {_.map(Array.prototype.slice.call(nodes), (node, index) => {
         let childNodes;
 
+        let sep = '.';
+        if (!parent || node.type.charAt(0) == '[') {
+          sep = '';
+        }
+
+        let position = (parent ? parent+sep : '') + node.type;
+
         if (typeof node.nodes !== 'undefined') {
           childNodes = <div className="TreeView__child">
-            <TreeView nodes={node.nodes} fetchedSigners={fetchedSigners}/>
+            <TreeView nodes={node.nodes} fetchedSigners={fetchedSigners} parent={position} />
           </div>;
         }
 
         return <div className="TreeView__set" key={index}>
           <div className="TreeView__row" key={index + node.type}>
-            <RowValue node={node} />
+            <RowValue node={node} position={position} />
           </div>
           {childNodes}
         </div>
@@ -67,7 +74,7 @@ function checkSignatures(nodes, fetchedSigners) {
 
 function RowValue(props) {
   let value, childNodes, separatorNeeded, separator;
-  let {node} = props;
+  let {node, position} = props;
 
   if (typeof node.value === 'string') {
     value = String(node.value);
@@ -95,7 +102,7 @@ function RowValue(props) {
     return formatSignature(node, separator);
   }
 
-  return <span><strong>{node.type}</strong>{separator}{value}</span>
+  return <span><strong>{node.type} {position}</strong>{separator}{value}</span>
 }
 
 // Types values are values that will be displayed with special formatting to
```
Then instead of:
```js
if(nodes[1] && nodes[1].type == "signatures")
```
We can check if we are in the exact location in a tree:
```js
if (position == 'TransactionEnvelope.signatures')
```